presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/CrossJoinWithOrFilterToInnerJoin.java

```diff
@@ -0,0 +1,274 @@
+                return true;
+            }
+        }
+        else if (expression instanceof SpecialFormExpression) {
+            SpecialFormExpression specialFormExpression = (SpecialFormExpression) expression;
+            if (specialFormExpression.getForm().equals(OR)) {
+                checkState(specialFormExpression.getArguments().size() == 2);
+                return getEqualExpressionInOrCondition(specialFormExpression.getArguments().get(0), equalExpression) &&
+                        getEqualExpressionInOrCondition(specialFormExpression.getArguments().get(1), equalExpression);
+            }
+        }
+        return false;
+    }
+
+    private RewrittenJoinInput rewriteJoinInput(List<VariableReferenceExpression> variablesInOrCondition, PlanNode joinInput, Type finalJoinKeyType, VariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator)
+    {
+        Map<VariableReferenceExpression, VariableReferenceExpression> castVariableMap = new HashMap<>();
+        Map<VariableReferenceExpression, RowExpression> castExpressionMap = new HashMap<>();
+        if (!variablesInOrCondition.stream().allMatch(x -> x.getType().equals(finalJoinKeyType))) {
+            // cast all to varchar type
+            for (int i = 0; i < variablesInOrCondition.size(); ++i) {
+                CallExpression castExpression = call("CAST", functionAndTypeManager.lookupCast(CastType.CAST, variablesInOrCondition.get(i).getType(), VARCHAR), VARCHAR, variablesInOrCondition.get(i));
+                VariableReferenceExpression castVariable = variableAllocator.newVariable(castExpression);
+                castVariableMap.put(variablesInOrCondition.get(i), castVariable);
+                castExpressionMap.put(castVariable, castExpression);
+            }
+        }
+
+        ImmutableList.Builder<RowExpression> constantsArgument = ImmutableList.builder();
+        for (int i = 0; i < variablesInOrCondition.size(); ++i) {
+            constantsArgument.add(constant((long) i, INTEGER));
+        }
+        CallExpression arrayConstruct = call(functionAndTypeManager, "array_constructor", new ArrayType(INTEGER), constantsArgument.build());
+
+        VariableReferenceExpression arrayVariable = variableAllocator.newVariable(arrayConstruct);
+        ImmutableMap.Builder<VariableReferenceExpression, RowExpression> projectAssignment = ImmutableMap.builder();
+        PlanNode project = PlannerUtils.addProjections(joinInput, idAllocator, projectAssignment.put(arrayVariable, arrayConstruct).putAll(castExpressionMap).build());
+        VariableReferenceExpression unnestVariable = variableAllocator.newVariable("field", INTEGER);
+        UnnestNode unnest = new UnnestNode(joinInput.getSourceLocation(), idAllocator.getNextId(), project, project.getOutputVariables().stream().filter(x -> !x.equals(arrayVariable)).collect(toImmutableList()), ImmutableMap.of(arrayVariable, ImmutableList.of(unnestVariable)), Optional.empty());
```

Comment:
Move the cross join unnest part to plannerutils as we have some other rewrites that can use it.

Reply:
This is a specific use case here, I think maybe not necessary to move to utils.